File: subtract a square.py
# Assignment 1
# Asmaa Ismael Heikal
# ID: 20210067
# Game 7 : Subtract a square


import math

start = int(input("enter the number of coins you have to play : "))


def play(coins):
    while coins != 0:

        # getting the input from the user
        print("Player 1 enter a \"squared\" non-zero number less than or equal ", coins, " : ", end="")
        x = int(input())

        # check if there is invalid input
        if x > coins or x == 0 or math.isqrt(x) != math.sqrt(x):
            print("Invalid input")
            continue

        coins -= x

        # check if the winner is player 1
        if coins == 0:
            print("Player 1 is the winner")
            break

        else:
            i = 0
            while True:
                if i % 2 != 0:
                    break

                print("Player 2 enter a \"squared\" non-zero number less than or equal ", coins, " : ", end="")
                x = int(input())
                if x > coins or x == 0 or math.isqrt(x) != math.sqrt(x):
                    print("Invalid input")
                    continue

                coins -= x
                i += 1
                # check if the winner is player 2
                if coins == 0:
                    print("Player 2 is the winner")
                    break

            # to break the biggest loop
            if coins == 0:
                break


play(start)
print("*Congratulations*")

File: subtract a square.cpp
//Assignment 1
//Asmaa Ismael Heikal
//ID: 20210067
//Game 7 : Subtract a square
//c++ version


#include<iostream>
#include <cmath>
using namespace std;
int main()
{
    int coins,x;
    float y;
    cout << "enter the number of coins you have to play : " ;
    cin >> coins;
    while(coins!=0)
    {
        cout << "Player 1 enter a \"squared\" non-zero number less than or equal "<<coins <<": ";
        cin >>x;
        y=sqrt(x);
        //check if there is invalid input
        if(x>coins || x==0 || y!=(int)y )
        {
            cout << "Invalid input"<<endl;
            continue;
        }
        coins -= x;
        //check if the winner is player 1
        if(coins==0)
        {
            cout << "Player 1 is the winner" << endl;
            break;
        }
        else
        {
            int i=0;
            while (true)
            {
                if(i%2!=0)
                {
                    break;
                }
                cout << "Player 2 enter a \"squared\" non-zero number less than or equal "<<coins <<": ";
                cin >>x;
                y=sqrt(x);
                if(x>coins || x==0 || y!=(int)y )
                {
                    cout << "Invalid input"<<endl;
                    continue;
                }
                coins -= x;
                i++;
                //check if the winner is player 2
                if(coins==0)
                {
                    cout << "Player 2 is the winner"<<endl;
                    break;
                }
            }
            //to break the biggest loop
            if(coins==0)
            {
                break;
            }
        }
    }
    cout << "*Congratulations*" <<endl;
}
